tracer/test/Datadog.Trace.ClrProfiler.IntegrationTests/DynamicConfigurationTests.cs

```diff
@@ -0,0 +1,257 @@
+        {
+            using var agent = EnvironmentHelper.GetMockAgent(useTelemetry: true);
+            var processName = EnvironmentHelper.IsCoreClr() ? "dotnet" : "Samples.Console";
+            using var logEntryWatcher = new LogEntryWatcher($"{LogFileNamePrefix}{processName}*");
+
+            EnvironmentHelper.CustomEnvironmentVariables["DD_TRACE_SAMPLE_RATE"] = "0.9";
+            using var sample = StartSample(agent, string.Empty, string.Empty, aspNetCorePort: 5000);
+
+            try
+            {
+                _ = await logEntryWatcher.WaitForLogEntry(DiagnosticLog);
+
+                await UpdateAndValidateConfig(
+                    agent,
+                    logEntryWatcher,
+                    new Config
+                    {
+                        TraceSampleRate = .5,
+                    });
+
+                await UpdateAndValidateConfig(
+                    agent,
+                    logEntryWatcher,
+                    config: new Config(),
+                    expectedConfig: new Config
+                    {
+                        TraceSampleRate = .9 // When clearing the key from dynamic configuration, it should revert back to the initial value
+                    });
+            }
+            finally
+            {
+                if (!sample.HasExited)
+                {
+                    sample.Kill();
+                }
+            }
+        }
+
+        private static bool IsConfigurationChangedEvent(object obj)
+        {
+            return ((TelemetryWrapper)obj).IsRequestType(TelemetryRequestTypes.AppClientConfigurationChanged);
+        }
+
+        private async Task UpdateAndValidateConfig(MockTracerAgent agent, LogEntryWatcher logEntryWatcher, Config config, Config expectedConfig = null)
+        {
+            const string diagnosticLogRegex = @".+ (?<diagnosticLog>\{.+\})\s+(?<context>\{.+\})";
+
+            expectedConfig ??= config;
+
+            var fileId = Guid.NewGuid().ToString();
+
+            var request = await agent.SetupRcmAndWait(Output, new[] { ((object)new { lib_config = config }, DynamicConfigurationManager.ProductName, fileId) });
+
+            request.Client.State.ConfigStates.Should().ContainSingle(f => f.Id == fileId)
+               .Subject.ApplyState.Should().Be(ApplyStates.ACKNOWLEDGED);
+
+            var log = await logEntryWatcher.WaitForLogEntry(DiagnosticLog);
+
+            agent.WaitForLatestTelemetry(IsConfigurationChangedEvent);
+
+            AssertConfigurationChanged(agent.Telemetry, config);
+
+            using var context = new AssertionScope();
+            context.AddReportable("log", log);
+
+            var match = Regex.Match(log, diagnosticLogRegex);
+
+            match.Success.Should().BeTrue();
+
+            var json = JObject.Parse(match.Groups["diagnosticLog"].Value);
+
+            // Can't validate tracing_header_tags and tracing_service_mapping because they're not in the diagnostic log
```

Comment:
I was thinking about the diagnostic log again - I wonder if when we're in debug mode we should fully dump the config telemetry to the log (only in debug mode, because there'll be a lot of it...)? 🤔